core/src/main/java/org/apache/calcite/sql/SqlDialect.java

```diff
@@ -757,16 +780,19 @@ public SqlDialect getDialect() {
   private static class ContextImpl implements Context {
     private final DatabaseProduct databaseProduct;
     private final String databaseProductName;
-    private final String databaseVersion;
+    private final int databaseMajorVersion;
+    private final int databaseMinorVersion;
     private final String identifierQuoteString;
     private final NullCollation nullCollation;
 
     private ContextImpl(DatabaseProduct databaseProduct,
-        String databaseProductName, String databaseVersion,
-        String identifierQuoteString, NullCollation nullCollation) {
+                        String databaseProductName,
+                        Integer databaseMajorVersion, Integer databaseMinorVersion,
+                        String identifierQuoteString, NullCollation nullCollation) {
       this.databaseProduct = Preconditions.checkNotNull(databaseProduct);
```

Comment:
I think the new parameters should be `int` not `Integer`.

Also, please remove the added whitespace; house style is to indent parameters just 4 spaces.

Reply:
Also, I think you should keep the old `databaseVersion` parameter, even as you add `databaseMinorVersion` and `databaseMajorVersion`.

Reply:
I think keeping the old databaseVersion parameter might be confusing. If there are 2 things representing the same information, and those 2 things are nullable, a user may be confused which of those things is available while he/she writes the code. This I say because the current contract of creating the "ContextImpl" allows me to create an object that may have null fields, such as a dialect of a database whose version is unknown.

For example, if inside a dialect, if i need to check whether the version is 2.3, i could do it in 2 ways, by checking the string or the int values. And since the fields are nullable, i would also have to checks nulls in both places, which would make the code very messy

Reply:
I disagree. databaseVersion, databaseMajorVersion and databaseMinorVersion all come from the JDBC driver and drivers seem to provide all of them correctly. I don't think the person writing the dialect will be "confused"; they can come up with a strategy that works for that particular DB, and that will be good enough and robust, I think. The more information the better; the only reason we don't use the DatabaseMetaData (which has lots of information) is because of memory leaks.